Plumb Blob partitioning pref to Blob Registry

sysrqb opened the pull request:
#### 0df98689c59d069de39d9ddb7e7fe403bf15070e
<pre>
Plumb Blob partitioning pref to Blob Registry
<a href="https://bugs.webkit.org/show_bug.cgi?id=251139">https://bugs.webkit.org/show_bug.cgi?id=251139</a>
rdar://104635569

Reviewed by Chris Dumez.

In this patch we introduce a new WebPreference that controls whether or not the
blob registry enables partitioning. The Network process (and NetworkSession, in
particular) do not know about preferences on all ports, so we add a new
NetworkSession creation parameter and IPC message that updates the
NetworkProcess with the current preference state. We perform this update on
webpage creation and destruction, and NetworkSession creation.

However, disabling the preference does not necessarily disable partitioning. If
any of the existing pages have partitioning enabled, then partitioning remains
enabled. Partitioning is disabled when all existing pages&apos; preferences have the
preference disabled.

Tested manually with MiniBrowser

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/network/BlobRegistryImpl.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setBlobRegistryTopOriginPartitioningEnabled const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_shouldAcceptInsecureCertificatesForWebSockets):
(WebKit::NetworkSession::setBlobRegistryTopOriginPartitioningEnabled):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setBlobRegistryTopOriginPartitioningEnabled):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::updateBlobRegistryPartitioningState const):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::isBlobRegistryPartitioningEnabled const):
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: <a href="https://commits.webkit.org/261783@main">https://commits.webkit.org/261783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c81b2e5350542cec244c083d79c500d605642989

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112741 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1437 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4543 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121284 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116811 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23264 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13088 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5711 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118510 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100524 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105848 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99244 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1069 "1 flakes 4 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46301 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/101047 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14235 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1110 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12362 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14927 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10456 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102556 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20252 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53101 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/32013 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8215 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16774 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110599 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27315 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/network/BlobRegistryImpl.h:
```diff
@@ -91,6 +92,7 @@ class WEBCORE_EXPORT BlobRegistryImpl {
     HashCountedSet<String> m_blobReferences;
     MemoryCompactRobinHoodHashMap<String, RefPtr<BlobData>> m_blobs;
     String m_fileDirectory;
+    bool m_partitionEnabled { false };
```
Boolean variables need a prefix, per coding style. Should probably be `m_isPartitioningEnabled`.

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.mm:
```diff
@@ -82,6 +82,7 @@
 }
 
 class WebBlobRegistry final : public BlobRegistry {
+    void setPartitioningEnabled(bool enabled) final { m_blobRegistry.setPartitioningEnabled(enabled); }
```
What is this doing here? Why isn't it below with the other member functions?

Almost looks as if you moved it out of the `private:` scope to make it public? However, it is still implicitly private so clearly this doesn't need to be public.

sysrqb replied:
I dropped the WebKitLegacy changes from the patch. The FIXME was a stale comment, but I'd rather concentrate on WebKit right now, and I can come back to WebKitLegacy, if needed. I'll resolve the other comments about it.

cdumez commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2934,12 +2935,16 @@ - (void)_preferencesChanged:(WebPreferences *)preferences
     settings.setRequiresUserGestureForVideoPlayback(mediaPlaybackRequiresUserGesture || [preferences videoPlaybackRequiresUserGesture]);
     settings.setRequiresUserGestureForAudioPlayback(mediaPlaybackRequiresUserGesture || [preferences audioPlaybackRequiresUserGesture]);
 
+    WTFLogAlways("WebView _preferencesChanged Setting webSQLEnabled as %d", [preferences webSQLEnabled]);
```
Should not be there.

cdumez commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2934,12 +2935,16 @@ - (void)_preferencesChanged:(WebPreferences *)preferences
     _private->page->setSessionID([preferences privateBrowsingEnabled] ? PAL::SessionID::legacyPrivateSessionID() : PAL::SessionID::defaultSessionID());
     _private->page->setBroadcastChannelRegistry(WebBroadcastChannelRegistry::getOrCreate([preferences privateBrowsingEnabled]));
     _private->group->storageNamespaceProvider().setSessionIDForTesting([preferences privateBrowsingEnabled] ? PAL::SessionID::legacyPrivateSessionID() : PAL::SessionID::defaultSessionID());
+    WTFLogAlways("WebView _preferencesChanged Setting blobRegistryTopOriginPartitioningEnabled as %d", [preferences blobRegistryTopOriginPartitioningEnabled]);
```
Should not be there.

cdumez commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2934,12 +2935,16 @@ - (void)_preferencesChanged:(WebPreferences *)preferences
     _private->group->storageNamespaceProvider().setSessionIDForTesting([preferences privateBrowsingEnabled] ? PAL::SessionID::legacyPrivateSessionID() : PAL::SessionID::defaultSessionID());
+    WTFLogAlways("WebView _preferencesChanged Setting blobRegistryTopOriginPartitioningEnabled as %d", [preferences blobRegistryTopOriginPartitioningEnabled]);
+    WebCore::platformStrategies()->blobRegistry()->setPartitioningEnabled([preferences blobRegistryTopOriginPartitioningEnabled]);
+    DeprecatedGlobalSettings::setWebSQLEnabled([preferences webSQLEnabled]);
```
I don't know that it is worth adding support for WebKitLegacy, especially if it means adding new code to DeprecatedGlobalSettings.

cdumez commented on Source/WebKitLegacy/win/WebCoreSupport/WebPlatformStrategies.cpp:
```diff
@@ -78,6 +78,7 @@ MediaStrategy* WebPlatformStrategies::createMediaStrategy()
 }
 
 class WebBlobRegistry final : public BlobRegistry {
+    void setPartitioningEnabled(bool enabled) final { m_blobRegistry.setPartitioningEnabled(enabled); }
```
Ditto. Weird location for this member function.

cdumez commented on Source/WebCore/page/DeprecatedGlobalSettings.h:
```diff
@@ -109,6 +109,9 @@ class DeprecatedGlobalSettings {
     static void setWebSQLEnabled(bool isEnabled) { shared().m_webSQLEnabled = isEnabled; }
     static bool webSQLEnabled() { return shared().m_webSQLEnabled; }
 
+    static void setBlobRegistryTopOriginPartitioningEnabled(bool isEnabled) { shared().m_blobRegistryTopOriginPartitioningEnabled = isEnabled; }
```
We should probably not be adding any new code to DeprecatedGlobalSettings.h.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1709,6 +1709,11 @@ void WebsiteDataStore::isResourceLoadStatisticsEphemeral(CompletionHandler<void(
 #endif
 }
 
+void WebsiteDataStore::setBlobRegistryTopOriginPartitioningEnabled(bool enabled)
```
What calls this?

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1125,6 +1125,8 @@ Ref<WebPageProxy> WebProcessPool::createWebPage(PageClient& pageClient, Ref<API:
 
     auto page = process->createWebPage(pageClient, WTFMove(pageConfiguration));
 
+    page->websiteDataStore().networkProcess().setBlobRegistryTopOriginPartitioningEnabled(page->websiteDataStore().sessionID(), page->preferences().store().getBoolValueForKey(WebPreferencesKey::blobRegistryTopOriginPartitioningEnabledKey()));
```
This isn't great. If you create one page with the setting on, then another with the setting off, you end up with partitioning disabled.

Ideally, we'd turn it on as soon as one page requires it and turn it off as soon as no page requires it.
We could have an `updateBlobRegistryPartitioningState()` function that'd we'd call every time a page gets added or removed. This function would iterate through all pages and see if any page needs the setting to be on.

sysrqb replied:
I was more concerned with testing, and so immediately enabling/disabling partitioning based on the preference seemed like the most obvious behavior, but I'm fine with making it behave like you described.

cdumez reviewed (commented):
Most of the code in this patch seems to be about adding support for WebKitLegacy but not quite since I don't think it works based on the FIXME comments you added. Not totally convinced it is worth it.

weinig reviewed (changes requested):
Please do not add additional options to DeprecatedGlobalSettings.h.

cdumez commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -496,6 +496,12 @@ void NetworkSession::firePrivateClickMeasurementTimerImmediatelyForTesting()
 
+void NetworkSession::setBlobRegistryTopOriginPartitioningEnabled(bool enabled)
+{
+    RELEASE_LOG(Storage, "NetworkSession::setBlobRegistryTopOriginPartitioningEnabled as %s", enabled ? "enabled" : "disabled");
```
You shouldn't use `%s` in RELEASE_LOG as it will get stripped. You'll need to use:
```
RELEASE_LOG(Storage, "NetworkSession::setBlobRegistryTopOriginPartitioningEnabled as %" PUBLIC_LOG_STRING, enabled ? "enabled" : "disabled");
```

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1384,6 +1384,14 @@ void NetworkProcessProxy::setPrivateClickMeasurementDebugMode(PAL::SessionID ses
 
+void NetworkProcessProxy::setBlobRegistryTopOriginPartitioningEnabled(PAL::SessionID sessionID, bool enabled)
+{
+    if (!canSendMessage())
```
Not needed.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -671,9 +671,10 @@ void WebProcessProxy::setThirdPartyCookieBlockingMode(ThirdPartyCookieBlockingMo
 {
-    Ref<WebPageProxy> webPage = WebPageProxy::create(pageClient, *this, WTFMove(pageConfiguration));
+    auto isBlobPartitioningEnabled = isBlobRegistryPartitioningEnabled(*pageConfiguration->preferences());
+    Ref<WebPageProxy> webPage = WebPageProxy::create(pageClient, *this, WTFMove(pageConfiguration), isBlobPartitioningEnabled);
```
We shouldn't be passing `isBlobPartitioningEnabled` to `WebPageProxy::create()` I think. We already paid the pageConfiguration and isBlobPartitioningEnabled comes from the pageConfiguration.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -671,9 +671,10 @@ void WebProcessProxy::setThirdPartyCookieBlockingMode(ThirdPartyCookieBlockingMo
+    Ref<WebPageProxy> webPage = WebPageProxy::create(pageClient, *this, WTFMove(pageConfiguration), isBlobPartitioningEnabled);
 
-    addExistingWebPage(webPage.get(), BeginsUsingDataStore::Yes);
+    addExistingWebPage(webPage.get(), BeginsUsingDataStore::Yes, isBlobPartitioningEnabled);
```
Shouldn't be needed. We already pass the page and we can get the preferences from the page already.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1837,6 +1848,19 @@ void WebProcessProxy::updateBackgroundResponsivenessTimer()
 
+void WebProcessProxy::updateBlobRegistryPartitioningState(bool isBlobPartitioningEnabled) const
+{
+    HashSet<PAL::SessionID> alreadySentToSessions;
```
This whole logic is not needed. A WebProcess can only have a single session. Just use `WebProcessProxy::sessionID()`.

cdumez reviewed (approved):
LGTM with the change requested.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1837,6 +1848,14 @@ void WebProcessProxy::updateBackgroundResponsivenessTimer()
+void WebProcessProxy::updateBlobRegistryPartitioningState(bool isBlobPartitioningEnabled) const
+{
+    WTF::forEach(globalPageMap(), [isBlobPartitioningEnabled](auto& page) {
+        if (auto networkProcess = page.value ? page.value->websiteDataStore().networkProcessIfExists() : nullptr)
```
Not needed, the WebProcessProxy has a single network process.

cdumez commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -177,6 +177,8 @@ NetworkSession::NetworkSession(NetworkProcess& networkProcess, const NetworkSess
     setTrackingPreventionEnabled(parameters.resourceLoadStatisticsParameters.enabled);
 #endif
 
+    setBlobRegistryTopOriginPartitioningEnabled(parameters.blobRegistryTopOriginPartitioningEnabled);
```
This needs a 'is' prefix, per WebKit coding style (booleans need a prefix):
`parameters.isBlobRegistryTopOriginPartitioningEnabled `

cdumez commented on Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
```diff
@@ -491,6 +497,7 @@ std::optional<NetworkSessionCreationParameters> NetworkSessionCreationParameters
 #if !HAVE(NSURLSESSION_WEBSOCKET)
         , WTFMove(*shouldAcceptInsecureCertificatesForWebSockets)
 #endif
+        , WTFMove(*blobRegistryTopOriginPartitioningEnabled)
```
No WTFMove() for booleans.

cdumez commented on Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
```diff
@@ -115,6 +115,7 @@ struct NetworkSessionCreationParameters {
 #if !HAVE(NSURLSESSION_WEBSOCKET)
     bool shouldAcceptInsecureCertificatesForWebSockets { false };
 #endif
+    bool blobRegistryTopOriginPartitioningEnabled { false };
```
Needs 'is' prefix.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1907,6 +1907,7 @@ WebsiteDataStoreParameters WebsiteDataStore::parameters()
 #if !HAVE(NSURLSESSION_WEBSOCKET)
     networkSessionParameters.shouldAcceptInsecureCertificatesForWebSockets = m_configuration->shouldAcceptInsecureCertificatesForWebSockets();
 #endif
+    networkSessionParameters.blobRegistryTopOriginPartitioningEnabled = WebProcessProxy::isBlobRegistryPartitioningEnabled();
```
That's not quite right, we only want to iterate over the pages that are associated with this data store, not ALL the pages.

sysrqb replied:
Right, good point, and in the other case (when a webpage is added/removed), do we want to look at all pages or only pages in that same web process or all pages associated with the same datastore? It seems like we want similar behavior in both places. I believe in both cases we want to look at all pages that are associated with the same data store. Does that make sense and do you agree?

cdumez replied:
Yes, this is what I suggested.

sysrqb replied:
Sorry, I didn't see your last review until after I pushed a new version and asked this question. Your review comments were helpful, thanks.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1907,6 +1907,13 @@ WebsiteDataStoreParameters WebsiteDataStore::parameters()
+    networkSessionParameters.isBlobRegistryTopOriginPartitioningEnabled = [this](){
+        bool isBlobPartitioningEnabled { false };
+        WebProcessProxy::forWebPagesWithDataStore(*this, [&isBlobPartitioningEnabled](const WebPageProxy& page) {
+            isBlobPartitioningEnabled |= page.preferences().blobRegistryTopOriginPartitioningEnabled();
```
Should use `||=` since we're not dealing with bits here.

That said, this should call `WebProcessProxy::isBlobRegistryPartitioningEnabled(sessionID)` to avoid code duplication.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -960,6 +972,13 @@ bool WebProcessProxy::shouldAllowNonValidInjectedCode() const
 }
 #endif
 
+bool WebProcessProxy::isBlobRegistryPartitioningEnabled()
```
It is never accurate to use ALL webpages to determine a setting when the setting is per session. This function should be taking a sessionID and only iterate through pages / processes with the given sessionID.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -717,6 +726,7 @@ void WebProcessProxy::addExistingWebPage(WebPageProxy& webPage, BeginsUsingDataS
 
     updateRegistrationWithDataStore();
     updateBackgroundResponsivenessTimer();
+    updateBlobRegistryPartitioningState(isBlobRegistryPartitioningEnabled());
```
Need to pass the sessionID()

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -748,6 +758,8 @@ void WebProcessProxy::removeWebPage(WebPageProxy& webPage, EndsUsingDataStore en
     updateMediaStreamingActivity();
     updateBackgroundResponsivenessTimer();
 
+    updateBlobRegistryPartitioningState(isBlobRegistryPartitioningEnabled());
```
ditto

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -197,6 +197,15 @@ void WebProcessProxy::forWebPagesWithOrigin(PAL::SessionID sessionID, const Secu
     }
 }
 
+void WebProcessProxy::forWebPagesWithDataStore(const WebsiteDataStore& websiteDataStore, const Function<void(WebPageProxy&)>& callback)
```
This can be folded into `WebProcessProxy::isBlobRegistryPartitioningEnabled(SessionID)`.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -197,6 +197,15 @@ void WebProcessProxy::forWebPagesWithOrigin(PAL::SessionID sessionID, const Secu
+void WebProcessProxy::forWebPagesWithDataStore(const WebsiteDataStore& websiteDataStore, const Function<void(WebPageProxy&)>& callback)
+{
+    for (auto& page : globalPages()) {
+        if (!page || &page->websiteDataStore() != &websiteDataStore)
```
I'd prefer if we compared sessionID's instead of data store pointers. It is identical at the moment but I find it clearer.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -717,6 +717,8 @@ void WebProcessProxy::addExistingWebPage(WebPageProxy& webPage, BeginsUsingDataS
 
     updateRegistrationWithDataStore();
     updateBackgroundResponsivenessTimer();
+    if (websiteDataStore())
```
I think we can avoid a little bit of code duplication by just calling this here:
`updateBlobRegistryPartitioningState();`

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -748,6 +750,9 @@ void WebProcessProxy::removeWebPage(WebPageProxy& webPage, EndsUsingDataStore en
     updateMediaStreamingActivity();
     updateBackgroundResponsivenessTimer();
 
+    if (websiteDataStore())
```
Just `updateBlobRegistryPartitioningState();` here as well.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1837,6 +1842,12 @@ void WebProcessProxy::updateBackgroundResponsivenessTimer()
     m_backgroundResponsivenessTimer.updateState();
 }
 
+void WebProcessProxy::updateBlobRegistryPartitioningState(bool isBlobPartitioningEnabled) const
```
Then we drop the parameter here and do:
```
void WebProcessProxy::updateBlobRegistryPartitioningState()
{
    auto* dataStore = websiteDataStore();
    if (auto* networkProcess = dataStore ? dataStore->networkProcessIfExists() : nullptr)
        networkProcess->setBlobRegistryTopOriginPartitioningEnabled(sessionID(), dataStore->isBlobPartitioningEnabled());
}
```

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1781,6 +1781,17 @@ void WebsiteDataStore::clearResourceLoadStatisticsInWebProcesses(CompletionHandl
+bool WebsiteDataStore::isBlobRegistryPartitioningEnabled() const
+{
+    return WTF::anyOf(m_processes, [this] (const WebProcessProxy& process) {
+        ASSERT(process.websiteDataStore());
```
I don't think those 2 assertions are needed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1781,6 +1781,17 @@ void WebsiteDataStore::clearResourceLoadStatisticsInWebProcesses(CompletionHandl
 
+bool WebsiteDataStore::isBlobRegistryPartitioningEnabled() const
+{
+    return WTF::anyOf(m_processes, [this] (const WebProcessProxy& process) {
```
I suspect we only need to capture `this` because of the assertions?

cdumez reviewed (approved):
This looks correct now. I made a few cosmetic suggestions.

sysrqb committed:
Plumb Blob partitioning pref to Blob Registry
https://bugs.webkit.org/show_bug.cgi?id=251139
rdar://104635569

Reviewed by Chris Dumez.

In this patch we introduce a new WebPreference that controls whether or not the
blob registry enables partitioning. The Network process (and NetworkSession, in
particular) do not know about preferences on all ports, so we add a new
NetworkSession creation parameter and IPC message that updates the
NetworkProcess with the current preference state. We perform this update on
webpage creation and destruction, and NetworkSession creation.

However, disabling the preference does not necessarily disable partitioning. If
any of the existing pages have partitioning enabled, then partitioning remains
enabled. Partitioning is disabled when all existing pages' preferences have the
preference disabled.

Tested manually with MiniBrowser

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/network/BlobRegistryImpl.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setBlobRegistryTopOriginPartitioningEnabled const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_shouldAcceptInsecureCertificatesForWebSockets):
(WebKit::NetworkSession::setBlobRegistryTopOriginPartitioningEnabled):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setBlobRegistryTopOriginPartitioningEnabled):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::updateBlobRegistryPartitioningState const):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::isBlobRegistryPartitioningEnabled const):
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: https://commits.webkit.org/261783@main
